### PR TITLE
Upgrade to ghc 8.10.3 in GitHub actions CI

### PR DESCRIPTION
**Pull request by newhoggy**
This PR is on top of https://github.com/input-output-hk/ouroboros-network/pull/2806

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -30,185 +34,295 @@ jobs:
...
         echo "LIBSODIUM_PATH=$LIBSODIUM_PATH"
         echo "$LIBSODIUM_PATH" >> $GITHUB_PATH
 
-    - name: Install libsodium
+    - name: Install libsodium (MacOS)
       if: matrix.os == 'macos-latest'
       run: brew install libsodium
 
-    - name: Configure to use libsodium
+    - name: Install build environment
```
Comment (coot) [inline .github/workflows/build.yml:77]: I'd call it `Install libsodium (Linux)` to be consistent with other similar steps.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -30,185 +34,295 @@ jobs:
...
+    - uses: actions/cache@v2
+      name: Cache cabal store
+      with:
+        path: ${{ steps.setup-haskell.outputs.cabal-store }}
+        key: cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+        restore-keys: |
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('dependencies.txt') }}
+          cache-${{ env.CACHE_VERSION }}-${{ runner.os }}-${{ matrix.ghc }}-
+
+    - name: Use cabal.project.local.ci.windows
```
Comment (coot) [inline .github/workflows/build.yml:131]: Shouldn't we use this only on `Windows` as the name `cabal.project.local.ci.windows` suggests?

File: .github/workflows/build.yml (hunk 1 above)
Comment (newhoggy) [reply .github/workflows/build.yml]: Done.  Thanks!

File: .github/workflows/build.yml (hunk 2 above)
Comment (newhoggy) [reply .github/workflows/build.yml]: Fixed.  Thanks!

**Comment by nfrisby** [comment]
I'm checking locally how difficult it'd be to avoid `-Wno-unused-matches`. Otherwise seems good, thanks 👍 .

**Comment by nfrisby** [comment]
Ryan Scott has already patched `bifunctors` so that the `-unused-match` warnings are no longer generated by the splices added in this PR.

I don't know when `bifunctors` will cut a new release with that, and even if I did, I don't know when -- or even admittedly how -- we'd adopt that. But eventually we don't need to disable those warnings anymore 🙌.

(Well, I'm pretty sure we won't need the warnings. Like I said, I'm not sure how to even temporarily test our code with a newer `bifunctors` --- since I only use the nix env and I haven't earned much mastery of `haskell.nix` yet.)

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -121,9 +121,25 @@ jobs:
     - name: Cabal Configure
       run: cabal --builddir="$CABAL_BUILDDIR" configure --enable-tests
 
-    - name: Use cabal.project.local.ci.windows
+    - name: Record dependencies
       run: |
-        cat ./cabal.project.local.ci.windows >> ./cabal.project.local
```
Comment (coot) [inline .github/workflows/build.yml:126]: How do we set `-Werror` option, this used to be done in `cabal.project.local.ci.windows`?

Comment (newhoggy) [reply .github/workflows/build.yml:126]: The diff is a bit funny.  The `cat ./cabal.project.local.ci.windows >> ./cabal.project.local` is preserved, but is lower down.  I've added an `if` statement around it so it is only used for Windows.

**Pull request closed without merging**